File: Packages/Yage/Tests/YageTests/AnimationFrameTests.swift
import XCTest

@testable import Yage

class AnimationFrameTests: XCTestCase {
    var entity: Entity!

    override func setUp() {
        entity = Entity(
            species: .agent,
            id: "test",
            frame: CGRect(x: 5, y: 5, width: 5, height: 5),
            in: World(name: "", bounds: CGRect(x: 0, y: 0, width: 25, height: 25))
        )
    }

    func testAnimationFrameIsSameAsEntityFrameIfNoCustomSizeOrPositionAreSet() {
        let animation = EntityAnimation(
            id: "test",
            size: nil,
            position: .entityTopLeft,
            facingDirection: nil
        )
        XCTAssertEqual(entity.frame, animation.frame(for: entity))
    }

    func testAnimationFrameIsCorrectWhenLargerThanEntityFromBottomLeft() {
        let animation = EntityAnimation(
            id: "test",
            size: CGSize(width: 2, height: 2),
            position: .fromEntityBottomLeft,
            facingDirection: nil
        )
        let expected = CGRect(x: 5, y: 0, width: 10, height: 10)
        let animationFrame = animation.frame(for: entity)
        XCTAssertEqual(expected, animationFrame)
        XCTAssertEqual(expected.bottomLeft, animationFrame.bottomLeft)
        XCTAssertEqual(entity.frame.bottomLeft, animationFrame.bottomLeft)
    }
}

File: Sources/windows/yage/tests/animation_frame_tests.py
import unittest

from yage.models.animations import EntityAnimation, EntityAnimationPosition
from yage.models.entity import Entity
from yage.utils.geometry import Rect, Size
from yage.models.species import SPECIES_AGENT
from yage.models.world import World


class AnimationFrameTests(unittest.TestCase):
    def setUp(self):
        self.entity = Entity(
            SPECIES_AGENT,
            "test",
            Rect(5, 5, 5, 5),
            World('', Rect(0, 0, 15,  15))
        )

    def test_animation_frame_is_same_as_entity_frame_if_no_custom_size_or_position_are_set(self):
        animation = EntityAnimation(
            "test",
            None,
            EntityAnimationPosition.ENTITY_TOP_LEFT,
            None
        )
        self.assertEqual(self.entity.frame, animation.frame(self.entity))

    def test_animation_frame_is_correct_when_larger_than_entity_from_bottom_left(self):
        animation = EntityAnimation(
            "test",
            Size(width=2, height=2),
            EntityAnimationPosition.FROM_ENTITY_BOTTOM_LEFT,
            None
        )
        expected = Rect(x=5, y=0, width=10, height=10)
        animation_frame = animation.frame(self.entity)
        self.assertEqual(expected, animation_frame)
        self.assertEqual(expected.bottom_left, animation_frame.bottom_left)
        self.assertEqual(self.entity.frame.bottom_left,
                         animation_frame.bottom_left)


if __name__ == '__main__':
    unittest.main()
